Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,326 @@
+
+* Aₙ, Bₙ, Dₙ, I₂ₘ
+
+And the six exceptional systems:
+
+* E₆, E₇, E₈, F₄, H₃, H₄
+
+## Main definitions
+
+* `Matrix.IsCoxeter` : A matrix `IsCoxeter` if it is a symmetric matrix with diagonal entries
+  equal to one and off-diagonal entries distinct from one.
+* `Matrix.CoxeterGroup` : The group presentation corresponding to a Coxeter matrix.
+* `CoxeterSystem` : A structure recording the isomorphism between a group `W` and the group
+  presentation corresponding to a Coxeter matrix, i.e. `Matrix.CoxeterGroup M`.
+* `IsCoxeterGroup` : A group is a Coxeter group if it is registered in a Coxeter system.
+
+## References
+
+* https://en.wikipedia.org/wiki/Coxeter_group
+
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 4--6*](bourbaki1968) chapter IV
+  pages 13--15
+
+## Tags
+
+coxeter system, coxeter group
+-/
+
+noncomputable section
+
+variable {B : Type*} [DecidableEq B]
+
+variable (M : Matrix B B ℕ)
+
+/-- A matrix `IsCoxeter` if it is a symmetric matrix with diagonal entries equal to one
+and off-diagonal entries distinct from one. -/
+class Matrix.IsCoxeter (M : Matrix B B ℕ) : Prop where
+  symmetric : M.IsSymm := by aesop
+  diagonal : ∀ i : B, M i i = 1 := by aesop
+  off_diagonal : ∀ i j : B, i ≠ j → M i j ≠ 1 := by aesop
+
+namespace CoxeterGroup
+
+namespace Relations
+
+/-- The relations corresponding to a Coxeter matrix. -/
+@[nolint unusedArguments]
+def ofMatrix [M.IsCoxeter] : B × B → FreeGroup B :=
+  Function.uncurry fun i j =>
+    (FreeGroup.of i * FreeGroup.of j) ^ M i j
+
+/-- The set of relations corresponding to a Coxeter matrix. -/
+def toSet [M.IsCoxeter] : Set (FreeGroup B) :=
+  Set.range <| ofMatrix M
+
+end Relations
+
+end CoxeterGroup
+
+/-- The group presentation corresponding to a Coxeter matrix. -/
+abbrev Matrix.CoxeterGroup [M.IsCoxeter] := PresentedGroup <| CoxeterGroup.Relations.toSet M
+
+/-- A Coxeter system `CoxeterSystem W` is a structure recording the isomorphism between
+a group `W` and the group presentation corresponding to a Coxeter matrix. Equivalently, this can
+be seen as a list of generators of `W` parameterized by the underlying type of `M`, which
+satisfy the relations of the Coxeter matrix `M`. -/
+structure CoxeterSystem [M.IsCoxeter] (W : Type*) [Group W]  where
+  /-- `CoxeterSystem.ofMulEquiv` constructs a Coxeter system given an equivalence with the group
+  presentation corresponding to a Coxeter matrix `M`. -/
+  ofMulEquiv ::
+    /-- `mulEquiv` is the isomorphism between the group `W` and the group presentation
+    corresponding to a Coxeter matrix `M`. -/
+    mulEquiv : Matrix.CoxeterGroup M ≃* W
+
+/-- Coxeter system of type `M` on the group `M.CoxeterGroup`. -/
+def Matrix.coxeterSystemCoxeterGroup [M.IsCoxeter] : CoxeterSystem M M.CoxeterGroup :=
+  CoxeterSystem.ofMulEquiv (MulEquiv.refl (CoxeterGroup M))
+
+/-- A group is a Coxeter group if it is registered in a Coxeter system. -/
+class IsCoxeterGroup [M.IsCoxeter] (W : Type*) [Group W] : Prop where
```

Comment:
This definition should rather be: there exists `B : Type`, there exists `M : Matrix B B ℕ`, such that `IsCoxeter M` and `Nonempty (CoxeterSystem M W)`.

Reply:
If you look back in the commit history I had what you are stating except for the existential quantifier on `B`.  It was then suggested to me in a previous code review that we should be having `[M.IsCoxeter]` for the definitions, so that is why I moved things around.  What is the consensus from the reviewers on this?

Reply:
No, I did the exact same suggestion as @sgouezel!

Reply:
I was referring to the [`M.IsCoxeter]`.  I wasn't implying that you were not suggesting the same thing with `∃ B`, because you did suggest that as well.  I shouldn't have bundled the comments together as I did, since it caused some miscommunication on my part.